flink/src/main/java/org/apache/iceberg/flink/FlinkCatalog.java

```diff
@@ -0,0 +1,508 @@
+ * would supply the first level in the catalog configuration and the second level would be exposed as Flink databases.
+ * <p>
+ * The Iceberg table manages its partitions by itself. The partition of the Iceberg table is independent of the
+ * partition of Flink.
+ */
+public class FlinkCatalog extends AbstractCatalog {
+
+  private final Catalog originalCatalog;
+  private final Catalog icebergCatalog;
+  private final String[] baseNamespace;
+  private final SupportsNamespaces asNamespaceCatalog;
+
+  public FlinkCatalog(
+      String catalogName,
+      String defaultDatabase,
+      String[] baseNamespace,
+      Catalog icebergCatalog,
+      boolean cacheEnabled) {
+    super(catalogName, defaultDatabase);
+    this.originalCatalog = icebergCatalog;
+    this.icebergCatalog = cacheEnabled ? CachingCatalog.wrap(icebergCatalog) : icebergCatalog;
+    this.baseNamespace = baseNamespace;
+    if (icebergCatalog instanceof SupportsNamespaces) {
+      asNamespaceCatalog = (SupportsNamespaces) icebergCatalog;
+    } else {
+      asNamespaceCatalog = null;
+    }
+  }
+
+  @Override
+  public void open() throws CatalogException {
+  }
+
+  @Override
+  public void close() throws CatalogException {
+    if (originalCatalog instanceof Closeable) {
+      try {
+        ((Closeable) originalCatalog).close();
+      } catch (IOException e) {
+        throw new CatalogException(e);
+      }
+    }
+  }
+
+  private Namespace toNamespace(String database) {
+    String[] namespace = new String[baseNamespace.length + 1];
+    System.arraycopy(baseNamespace, 0, namespace, 0, baseNamespace.length);
+    namespace[baseNamespace.length] = database;
+    return Namespace.of(namespace);
+  }
+
+  private TableIdentifier toIdentifier(ObjectPath path) {
+    return TableIdentifier.of(toNamespace(path.getDatabaseName()), path.getObjectName());
+  }
+
+  @Override
+  public List<String> listDatabases() throws CatalogException {
+    if (asNamespaceCatalog == null) {
+      return Collections.singletonList(getDefaultDatabase());
+    }
+
+    return listAllNamespaces(Namespace.empty()).stream()
+        .map(n -> n.level(n.levels().length - 1))
+        .collect(Collectors.toList());
+  }
+
+  private List<Namespace> listAllNamespaces(Namespace namespace) {
+    if (asNamespaceCatalog == null) {
+      throw new RuntimeException("The asNamespaceCatalog should not be null.");
+    }
+
+    String[] levels = namespace.levels();
+    if (levels.length == baseNamespace.length + 1) {
+      return Collections.singletonList(namespace);
+    }
+    if (levels.length < baseNamespace.length + 1) {
+      for (int i = 0; i < levels.length; i++) {
+        if (!baseNamespace[i].equals(levels[i])) {
+          return Collections.emptyList();
+        }
+      }
+      List<Namespace> ret = new ArrayList<>();
+      asNamespaceCatalog.listNamespaces(namespace).forEach(n -> ret.addAll(listAllNamespaces(n)));
```

Comment:
Why is this recursive? It seems unnecessary.

Reply:
You are right, `asNamespaceCatalog.listNamespaces(Namespace.of(baseNamespace))` is enough.